cpp/src/arrow/record_batch.cc

```diff
@@ -432,4 +433,43 @@ RecordBatchReader::~RecordBatchReader() {
   ARROW_WARN_NOT_OK(this->Close(), "Implicitly called RecordBatchReader::Close failed");
 }
 
+Result<std::shared_ptr<RecordBatch>> ConcatenateRecordBatches(
+    const RecordBatchVector& batches, MemoryPool* pool) {
+  int64_t length = 0;
+  size_t n = batches.size();
+  if (n == 0) {
+    return Status::Invalid("Must pass at least one recordbatch");
+  }
+  if (n == 1) {
+    return batches[0];
+  }
+  int cols = batches[0]->num_columns();
+  auto schema = batches[0]->schema();
+  std::vector<std::shared_ptr<Array>> columns;
+  if (cols == 0) {
+    // special case: null batch, no data, just length
+    for (size_t i = 0; i < batches.size(); ++i) {
+      length += batches[i]->num_rows();
+    }
+  } else {
```

Comment:
I don't think you need a special case for zero columns:

```suggestion
  // special case: null batch, no data, just length
  for (size_t i = 0; i < batches.size(); ++i) {
    length += batches[i]->num_rows();
  }
```

Then the loop `for (int col = 0; col < cols; ++col)` will just never be entered if `cols == 0`

Reply:
no, My pseudocode is below, with the else branch
if (cols == 0) {
    // special case: null batch, no data, just length
    for (size_t i = 0; i < batches.size(); ++i) {
      length += batches[i]->num_rows();
    }
  } else {
   for (int col = 0; col < cols; ++col)
}

Reply:
The reason why the special case of column 0 is processed here is that if we want to implement count(*), then this RecordBatch does not store data, that is, only the length. This scenario is triggered when small batches need to be merged.

Reply:
Will the following change make any difference on the result?
```cpp
Result<std::shared_ptr<RecordBatch>> ConcatenateRecordBatches(
    const RecordBatchVector& batches, MemoryPool* pool) {
  int64_t length = 0;
  size_t n = batches.size();
  if (n == 0) {
    return Status::Invalid("Must pass at least one recordbatch");
  }
  int cols = batches[0]->num_columns();
  auto schema = batches[0]->schema();
  for (int i = 0; i < batches.size(); ++i) {
    length += batches[i]->num_rows();
    if (!schema->Equals(batches[i]->schema())) {
      return Status::Invalid(
          "Schema of RecordBatch index ", i, " is ", batches[i]->schema()->ToString(),
          ", which does not match index 0 recordbatch schema: ", schema->ToString());
    }
  }

  std::vector<std::shared_ptr<Array>> concatenated_columns;
  for (int col = 0; col < cols; ++col) {
    ArrayVector column_arrays;
    for (const auto & batch : batches) {
      column_arrays.emplace_back(std::move(batch->column(col)));
    }
    ARROW_ASSIGN_OR_RAISE(auto concatenated_column, Concatenate(column_arrays, pool))
    concatenated_columns.emplace_back(std::move(concatenated_column));
  }
  return RecordBatch::Make(std::move(schema), length, std::move(concatenated_columns));
}
```
I assumed:
1. Schema checking can be moved out of the nested loop.
2. The result length is just the sum of all batch lengths, no matter how many (or zero) columns there are.
3. Each column has the same length in a RecordBatch, so no checking of column length is needed here.
4. Variable names like `column_data`, `data`, `columns` are ambiguous so I changed them to more descriptive names.

Reply:
The code looks very refreshing, but it does not handle the case of cols = 0. In this case, cols = 0, length is not obtained.

For a batch, there may be cols = 0, the schema is empty, but there is length. This kind of batch is used to process count(*) and does not store the actual columns, only the number of rows.

Reply:
```cpp
for (int i = 0; i < batches.size(); ++i) {
    length += batches[i]->num_rows();
    ...
}
```
Doesn't this include the cols=0 case?

Reply:
Sorry, you are correct, it is handled here

Reply:
Thanks for the modification. This code modification has been submitted.
 note:
```
column_arrays.emplace_back(std::move(batch->column(col)));
```
Change to
``` 
column_arrays.emplace_back(batch->column(col));
```